src/libstore/meson.build

```diff
@@ -421,4 +421,9 @@ install_headers(headers, subdir : 'nix', preserve_path : true)
 
 libraries_private = []
 
+extra_pkg_config_variables = {
+  'localstatedir' : get_option('state-dir'),
```

Comment:
And here you call it localstatedir! :D 

Which makes me wonder why you don't use meson's own --localstatedir option (https://www.gnu.org/prep/standards/standards.html#Directory-Variables, even) and just set that value to /nix/var, perhaps in default_options.

https://github.com/NixOS/nix/blob/31f3f23ee6ce6cb7ab68ada1682aa2eb39764210/src/libstore/meson.build#L1-L13

Reply:
Oh hmm, I just took this from Lix's Meson, but maybe the Lix devs didn't realize one could override a built-in option default value. CC @Qyriad

Reply:
Perhaps:
```patch
diff --git a/src/libstore/meson.build b/src/libstore/meson.build
index d0cd7d3cb..b829de3f7 100644
--- a/src/libstore/meson.build
+++ b/src/libstore/meson.build
@@ -421,9 +421,9 @@ install_headers(headers, subdir : 'nix', preserve_path : true)
 
 libraries_private = []
 
-extra_pkg_config_variables = {
-  'localstatedir' : get_option('state-dir'),
-  'storedir' : get_option('store-dir'),
-}
+extra_pkg_config_variables = [
+  f'localstatedir=' + get_option('state-dir'),
+  f'storedir=' + get_option('store-dir'),
+]
 
 subdir('build-utils-meson/export')

```
We do the dict to string conversion ourselves and it doesn't look like a Path, thus bypassing that check.

Reply:
Trying @eli-schwartz's `localstatedir=/nix/var` suggestion.

Reply:
Still a problem. I minimized it and opened https://github.com/mesonbuild/meson/issues/13584 Meson options are a red herring -- it's declared dependency variables instead.

Reply:
My comment was more about how it felt a bit unusual to me to be using declared dependency variables at all here.

Reply:
@eli-schwartz Well, I am open to alternatives! Here's basically what I am thinking:

1. The paths are a property of the `libstore` library, and configurable, and needs to be conveyed downstream.

2. We can't just put it in a header, because we need it at configure time downstream, patting `localstatedir` and `storedir` in the pkg-config file seems more appropriate.

4. When we are doing the combined subprojects build, we don't use a pkg-config dep, and instead use an internal dep. So we get the variables from the internal dep instead.

Where does it go off the rails to you, conceptually? Each step seems fine to me, even if the end result is a bit funky.

Reply:
When you're doing the combined subprojects build, do different subprojects get built with different prefix/libdir/localstatedir?

Reply:
Fair question: Nope. Each project should either use the same `localstatedir`, or doesn't care about the variable at all.

Reply:
So perhaps for the purpose of this PR we can dodge the question of what declare_dependency is doing, and simply have projects that want this value look it up via get_option() instead of subp_dep.get_variable()...

Reply:
I did something along that lines. I still need the pkg-config variable for `storedir`, but I now don't set the `localstatedir` one on macOS, and I have an assert that it lines up with `get_option('localstatedir')`.